### ISSUE TITLE
Add save() and load() methods to knn-classifier in tfjs-models

### ISSUE THREAD

**Issue by dsmilkov** [tensorflow/tfjs#633]
We should add save() and load() methods to KnnClassifier. They can take the same url/path format as model.save().

See [discussion](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!msg/tfjs/3sXgLFIiHts/AF8NAd88EwAJ) for motivation and context.

Internally we can make an empty model with non-trainable weights and use the existing model.save() infrastructure to save it.

Long term, we should have a generic tf.save(), tf.load() method that can take an dict of tensor names to tensors and save them.

cc @tafsiri @caisq in case I missed something from our discussion.

**Comment by oreHGA** [comment]
Hi, I'd like to give this a shot!

**Comment by hpssjellis** [comment]
@oreHGA  Any update on this? I wold really like to use this saving ability for classifier.getClassifierDataset(). As far as I can tell it is just saving an array of 2D tensors. Anyone active on this?

**Comment by hpssjellis** [comment]
So my idea of making a fake model with multiple inputs to save each tensor from the knn-classifier does work. 

Basically it takes the output from the `classifier.getClassifierDataset()`

which returns an array of tensors, then sets up a model with inputs and a dense layer for each trained classifier. That model can be saved using `model.save()` The saved model can then be loaded from a website and converted back into the classifier using `classifier.setClassifierDataset()` once the array of tensors have been extracted from the fake model.


Reply if you are interested in the code. I am still working on making the code a bit more readable. I had to first understand multiple inputs and here is a [demo](https://hpssjellis.github.io/beginner-tensorflowjs-examples-in-javascript/advanced-keras/45keras-multi-input.html) for that

**Comment by seppestaes** [comment]
Hi, I would like to take a look at the code!

**Comment by hpssjellis** [comment]
Took me a while to find the code. Here is the [knn-classifier demo](https://hpssjellis.github.io/beginner-tensorflowjs-examples-in-javascript/advanced-keras/face/43-face-classifier-saving.html). Look for the buttons "save-classifier" and "load-classifier". You can just view-source or visit the github.

The github is [here](https://github.com/hpssjellis/beginner-tensorflowjs-examples-in-javascript/blob/2497f865997fe33bbe5bb9e2917d2748aecc4eee/advanced-keras/face/43-face-classifier-saving.html#L332-L387)

**Comment by leung85** [comment]
This is my simply implement for save an load knn and it works, hope can help anyone who need :)
 ```
save() {
    let dataset = this.classifier.getClassifierDataset()
    var datasetObj = {}
    Object.keys(dataset).forEach((key) => {
      let data = dataset[key].dataSync();
      // use Array.from() so when JSON.stringify() it covert to an array string e.g [0.1,-0.2...] 
      // instead of object e.g {0:"0.1", 1:"-0.2"...}
      datasetObj[key] = Array.from(data); 
    });
    let jsonStr = JSON.stringify(datasetObj)
    //can be change to other source
    localStorage.setItem("myData", jsonStr);
  }
```
```
  load() {
     //can be change to other source
    let dataset = localStorage.getItem("myData")
    let tensorObj = JSON.parse(dataset)
    //covert back to tensor
    Object.keys(tensorObj).forEach((key) => {
      tensorObj[key] = tf.tensor(tensorObj[key], [tensorObj[key].length / 1000, 1000])
    })
    this.classifier.setClassifierDataset(tensorObj);
  }
```

**Comment by oveddan** [comment]
I would find this useful as well.  Based on @leung85 and @hpssjellis examples I've created a typescipt and async version:

```typescript
import * as knnClassifier from "@tensorflow-models/knn-classifier";
import * as tf from '@tensorflow/tfjs';

type Dataset = {
  [classId: number]: tf.Tensor<tf.Rank.R2>
};

type DatasetObjectEntry = {
  classId: number,
  data: number[],
  shape: [number, number]
};

type DatasetObject = DatasetObjectEntry[];

async function toDatasetObject(dataset: Dataset): Promise<DatasetObject> {
  const result: DatasetObject = await Promise.all(
    Object.entries(dataset).map(async ([classId,value], index) => {
      const data = await value.data();

      return {
        classId: Number(classId),
        data: Array.from(data),
        shape: value.shape
      };
   })
  );

  return result;
};

function fromDatasetObject(datasetObject: DatasetObject): Dataset {
  return Object.entries(datasetObject).reduce((result: Dataset, [indexString, {data, shape}]) => {
    const tensor = tf.tensor2d(data, shape);
    const index = Number(indexString);

    result[index] = tensor;

    return result;
  }, {});

}

const storageKey = "knnClassifier";

async function saveClassifierInLocalStorate(classifier: knnClassifier.KNNClassifier) {
  const dataset = classifier.getClassifierDataset();
  const datasetOjb: DatasetObject = await toDatasetObject(dataset);
  const jsonStr = JSON.stringify(datasetOjb);
  //can be change to other source
  localStorage.setItem(storageKey, jsonStr);
}

function loadClassifierFromLocalStorage(): knnClassifier.KNNClassifier {
  const classifier: knnClassifier.KNNClassifier = new knnClassifier.KNNClassifier();

  const datasetJson = localStorage.getItem(storageKey);

  if (datasetJson) {
    const datasetObj = JSON.parse(datasetJson) as DatasetObject;

    const dataset = fromDatasetObject(datasetObj);

    classifier.setClassifierDataset(dataset);
  }
  return classifier;
}
```

**Comment by hpssjellis** [comment]
That's great @oveddan, any chance of that being sent as a PR to TFJS or is it too specific and should just be loaded as needed?

**Comment by jonnytest1** [comment]
@leung85 newer versions use tensor size of 1024

**Comment by josephrocca** [comment]
Maybe it's obvious, but you can read the shape from the tensor object, so no need to hard-code in 1000 or 1024, etc. Based on leung85's example (sorry for long lines):

```js
let classifier = knnClassifier.create();
// Add some examples:
classifier.addExample(...);
// Save it to a string:
let str = JSON.stringify( Object.entries(classifier.getClassifierDataset()).map(([label, data])=>[label, Array.from(data.dataSync()), data.shape]) );
// Load it back into a fresh classifier:
classifier = knnClassifier.create();
classifier.setClassifierDataset( Object.fromEntries( JSON.parse(str).map(([label, data, shape])=>[label, tf.tensor(data, shape)]) ) );
```

**Comment by VladimirHumeniuk** [comment]
> This is my simply implement for save an load knn and it works, hope can help anyone who need :)
> 
> ```
> save() {
>    let dataset = this.classifier.getClassifierDataset()
>    var datasetObj = {}
>    Object.keys(dataset).forEach((key) => {
>      let data = dataset[key].dataSync();
>      // use Array.from() so when JSON.stringify() it covert to an array string e.g [0.1,-0.2...] 
>      // instead of object e.g {0:"0.1", 1:"-0.2"...}
>      datasetObj[key] = Array.from(data); 
>    });
>    let jsonStr = JSON.stringify(datasetObj)
>    //can be change to other source
>    localStorage.setItem("myData", jsonStr);
>  }
> ```
> 
> ```
>   load() {
>      //can be change to other source
>     let dataset = localStorage.getItem("myData")
>     let tensorObj = JSON.parse(dataset)
>     //covert back to tensor
>     Object.keys(tensorObj).forEach((key) => {
>       tensorObj[key] = tf.tensor(tensorObj[key], [tensorObj[key].length / 1000, 1000])
>     })
>     this.classifier.setClassifierDataset(tensorObj);
>   }
> ```

It's a very nice answer, but I got a `tf-core.esm.js?45ef:17 Uncaught (in promise) Error: Tensor must have a shape comprised of positive integers but got shape [13.312,1000].` error when I'm trying to load my saved to local storage data. Can you please provide any advice with this?

**Comment by VladimirHumeniuk** [comment]
@oveddan I tried to do it as in your answer but after `fromDatasetObject` my data-set has `undefined` in `classIndex`. Could you please provide any advice on how to fix it?

**Comment by Mxlt** [comment]
> @oveddan I tried to do it as in your answer but after `fromDatasetObject` my data-set has `undefined` in `classIndex`. Could you please provide any advice on how to fix it?

@VladimirHumeniuk  Hi! Have you been able to fix this? I am having the same issue.

**Comment by VladimirHumeniuk** [comment]
@Mxlt I just used `label` instead of `classIndex`

**Comment by swimauger** [comment]
Noticed most of these answers are synchronous, which is potentially dangerous if you're expecting to unload a large dataset, so I have created a library for parsing and stringifying datasets of these types. If you are interested, take a look at [tensorset](https://npmjs.com/package/tensorset). There is some documentation on how to use it, it works similar to `JSON.stringify` and `JSON.parse`.

Here is an example of using `Tensorset` with the KNN-Classifier:

```JavaScript
const fs = require('fs').promises;
const Tensorset = require('tensorset');

(async () => {
    // Create a classifier, add your examples
    const originalClassifier = knnClassifier.create();
    originalClassifier.addExample(/*Some Example*/);

    // Stringify the dataset
    let dataset = Tensorset.stringify(originalClassifier.getClassifierDataset());
    
    // Save the dataset
    await fs.writeFile(/*File Name*/, dataset);
    
    // Load the dataset
    dataset = await fs.readFile(/*File Name*/);

    // Parse the dataset
    dataset = await Tensorset.parse(dataset);

    // Add to a new classifier
    const newClassifier = knnClassifier.create();
    newClassifier.setClassifierDataset(dataset);
})();
```

**Comment by JJwilkin** [comment]
@swimauger This is really great, thank you for contributing these two packages! I was wondering for the image-classifier package, does it use MobileNet under the hood for feature extraction and the package you provide simply wrapping the mobilenet + KnnClassifier so it can take a jpg/png image as input?

**Comment by swimauger** [comment]
@JJwilkin precisely!

**Comment by 408881465** [comment]
> Maybe it's obvious, but you can read the shape from the tensor object, so no need to hard-code in 1000 or 1024, etc. Based on leung85's example (sorry for long lines):
> 
> ```js
> // Create your classifier:
> let classifier = knnClassifier.create();
> // Add some examples:
> classifier.addExample(...);
> // Save it to a string:
> let str = JSON.stringify( Object.entries(classifier.getClassifierDataset()).map(([label, data])=>[label, Array.from(data.dataSync()), data.shape]) );
> // Load it back into a fresh classifier:
> classifier = knnClassifier.create();
> classifier.setClassifierDataset( Object.fromEntries( JSON.parse(str).map(([label, data, shape])=>[label, tf.tensor(data, shape)]) ) );
> ```

Thx. It works for me.
